### ISSUE TITLE
Recent commit breaks EVP_aes_128_cbc_hmac_sha256

### ISSUE THREAD

**Issue by guidovranken** [openssl/openssl#12528]
Broken since f29dbb08668318b84d7bca0bd63c585e0169545e

Before this commit, the following program runs OK, and after it, it fails to decrypt its own ciphertext.

I know this is a special TLS cipher not intended for ordinary use, but my fuzzer detected this and I wanted to check with you if this is the intended behavior.

@slontis 

```cpp
#include <vector>
#include <openssl/evp.h>

#define CF_CHECK_EQ(expr, res) if ( (expr) != (res) ) { goto end; }
#define CF_CHECK_NE(expr, res) if ( (expr) == (res) ) { goto end; }

std::vector<unsigned char> encrypt(const std::vector<unsigned char>& cleartext, const std::vector<unsigned char>& key, const std::vector<unsigned char>& iv) {
    std::vector<unsigned char> out(256);
    bool ok = false;

    const EVP_CIPHER* cipher = nullptr;
    EVP_CIPHER_CTX* ctx = nullptr;

    CF_CHECK_NE(cipher = EVP_aes_128_cbc_hmac_sha256(), nullptr);
    CF_CHECK_NE(ctx = EVP_CIPHER_CTX_new(), nullptr);
    CF_CHECK_EQ(EVP_EncryptInit_ex(ctx, cipher, nullptr, nullptr, nullptr), 1);
    CF_CHECK_EQ(EVP_EncryptInit_ex(ctx, nullptr, nullptr, key.data(), iv.data()), 1);
    {
        size_t outIdx = 0;
        int len = -1;

        CF_CHECK_EQ(EVP_EncryptUpdate(ctx, out.data() + outIdx, &len, cleartext.data(), cleartext.size()), 1);
        outIdx += len;

        CF_CHECK_EQ(EVP_EncryptFinal_ex(ctx, out.data() + outIdx, &len), 1);
        outIdx += len;

        out.resize(outIdx);
    }

    ok = true;

end:
    if ( !ok ) { printf("Encryption failed\n"); abort(); }

    EVP_CIPHER_CTX_free(ctx);
    
    return out;
}

std::vector<unsigned char> decrypt(const std::vector<unsigned char>& ciphertext, const std::vector<unsigned char>& key, const std::vector<unsigned char>& iv) {
    std::vector<unsigned char> out(256);
    bool ok = false;

    const EVP_CIPHER* cipher = nullptr;
    EVP_CIPHER_CTX* ctx = nullptr;

    CF_CHECK_NE(cipher = EVP_aes_128_cbc_hmac_sha256(), nullptr);
    CF_CHECK_NE(ctx = EVP_CIPHER_CTX_new(), nullptr);
    CF_CHECK_EQ(EVP_DecryptInit_ex(ctx, cipher, nullptr, nullptr, nullptr), 1);
    CF_CHECK_EQ(EVP_DecryptInit_ex(ctx, nullptr, nullptr, key.data(), iv.data()), 1);
    {
        size_t outIdx = 0;
        int len = -1;

        CF_CHECK_EQ(EVP_DecryptUpdate(ctx, out.data() + outIdx, &len, ciphertext.data(), ciphertext.size()), 1);
        outIdx += len;

        CF_CHECK_EQ(EVP_DecryptFinal_ex(ctx, out.data() + outIdx, &len), 1);
        outIdx += len;

        out.resize(outIdx);
    }

    ok = true;

end:
    if ( !ok ) { printf("Decryption failed\n"); abort(); }

    EVP_CIPHER_CTX_free(ctx);
    
    return out;
}

int main(void)
{
    const std::vector<unsigned char> cleartext = {0x65, 0x00, 0x00, 0x00, 0x00, 0x10, 0x00, 0x00, 0x00, 0x2c, 0x00, 0x00, 0xc3, 0x00, 0x00, 0x00};
    const std::vector<unsigned char> key = {0x00, 0x00, 0x21, 0x00, 0x32, 0x00, 0x00, 0x00, 0x3b, 0x0e, 0x30, 0x00, 0xef, 0x99, 0x2a, 0x23};
    const std::vector<unsigned char> iv = {0x00, 0x0a, 0x00, 0x00, 0x09, 0x00, 0xce, 0x00, 0x14, 0x05, 0x6a, 0x00, 0x00, 0x00, 0x00, 0xff};

    const auto ciphertext = encrypt(cleartext, key, iv);
    const auto cleartext2 = decrypt(ciphertext, key, iv);
    if ( cleartext != cleartext2 ) {
        printf("Incorrect decryption\n");
    }

    printf("All ok\n");

    return 0;
}
```

**Comment by slontis** [comment]
You are correct - it should only be used by tls..
The new code removes tlspad - so it is not going to work as a simple cipher anymore.
@mattcaswell what do you think?

**Comment by t8m** [comment]
One of the commits from #12288 also broke the boringssl test - SSL interoperability. Not sure if this particular commit caused it. @mattcaswell

**Comment by t8m** [comment]
Is this still broken?

**Comment by guidovranken** [comment]
The following program

```cpp
#include <vector>
#include <openssl/evp.h>

#define CF_CHECK_EQ(expr, res) if ( (expr) != (res) ) { goto end; }
#define CF_CHECK_NE(expr, res) if ( (expr) == (res) ) { goto end; }

std::vector<unsigned char> decrypt(const std::vector<unsigned char>& ciphertext, const std::vector<unsigned char>& key, const std::vector<unsigned char>& iv) {
    std::vector<unsigned char> out(256);

    const EVP_CIPHER* cipher = nullptr;
    EVP_CIPHER_CTX* ctx = nullptr;

    CF_CHECK_NE(cipher = EVP_aes_128_cbc_hmac_sha256(), nullptr);
    CF_CHECK_NE(ctx = EVP_CIPHER_CTX_new(), nullptr);
    CF_CHECK_EQ(EVP_DecryptInit_ex(ctx, cipher, nullptr, nullptr, nullptr), 1);
    CF_CHECK_EQ(EVP_DecryptInit_ex(ctx, nullptr, nullptr, key.data(), iv.data()), 1);
    {
        size_t outIdx = 0;
        int len = -1;

        CF_CHECK_EQ(EVP_DecryptUpdate(ctx, out.data() + outIdx, &len, ciphertext.data(), ciphertext.size()), 1);
        outIdx += len;

        CF_CHECK_EQ(EVP_DecryptFinal_ex(ctx, out.data() + outIdx, &len), 1);
        outIdx += len;

        out.resize(outIdx);
    }

end:

    EVP_CIPHER_CTX_free(ctx);
    
    return out;
}

int main(void)
{
    const std::vector<unsigned char> cleartext(16);
    const std::vector<unsigned char> key(16);
    const std::vector<unsigned char> iv(16);

    decrypt(cleartext, key, iv);

    return 0;
}
```

now aborts with

```
providers/implementations/ciphers/ciphercommon.c:442: OpenSSL internal error: Assertion failed: *outl >= (size_t)(out[inl - 1] + 1)
```

This is OSS-Fuzz issue 28959: https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=28959

**Comment by mattcaswell** [comment]
Fix in #13774
